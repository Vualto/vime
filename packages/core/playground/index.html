<!DOCTYPE html>
<html dir="ltr">

<head>
  <meta charset="UTF-8">
  <title>Playground | Vime</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no">

  <link rel="stylesheet" href="/src/themes/default.css">
  <link rel="stylesheet" href="/src/themes/light.css">
  <link rel="icon" type="image/png" href="/playground/favicon.ico">
  <script type="module" src="/dist/vime/vime.esm.js"></script>

  <style>
    html, body {
      width: 100%;
      height: 100%;
      margin: 0;
      box-sizing: border-box;
    }

    body {
      padding: 16px;
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
    }
  </style>
</head>
<body>
  <vm-playground></vm-playground>
</body>
</html>
