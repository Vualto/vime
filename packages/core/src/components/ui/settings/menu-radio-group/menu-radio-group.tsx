import {
  h, Component, Element, Prop, Watch, Event, EventEmitter, Listen,
} from '@stencil/core';
import { withComponentRegistry } from '../../../core/player/withComponentRegistry';

/**
 * @slot - Used to pass in radio buttons (`vm-menu-radio`).
 */
@Component({
  tag: 'vm-menu-radio-group',
  shadow: true,
})
export class MenuRadioGroup {
  @Element() el!: HTMLVmMenuRadioGroupElement;

  /**
   * The current value selected for this group.
   */
  @Prop({ mutable: true }) value?: string;

  @Watch('value')
  onValueChange() {
    this.findRadios().forEach((radio) => {
      radio.checked = (radio.value === this.value);
    });
  }

  /**
   * Emitted when a new radio button is selected for this group.
   */
  @Event() vmCheck!: EventEmitter<void>;

  constructor() {
    withComponentRegistry(this);
  }

  connectedCallback() {
    this.onValueChange();
  }

  @Listen('vmCheck')
  onSelectionChange(event: Event) {
    const radio = event.target as HTMLVmMenuRadioElement;
    this.value = radio.value;
  }

  private findRadios() {
    return this.el.querySelectorAll('vm-menu-radio');
  }

  render() {
    return (
      <slot />
    );
  }
}
