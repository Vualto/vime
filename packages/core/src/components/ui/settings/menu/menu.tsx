import {
  h, Component, Prop, Listen, Event, EventEmitter, Element, State, Watch, Method, writeTask,
} from '@stencil/core';
import { buildNoAncestorSelector } from '../../../../utils/dom';
import { isUndefined, isNull } from '../../../../utils/unit';
import { withComponentRegistry } from '../../../core/player/withComponentRegistry';

/**
 * @slot - Used to pass in the body of the menu which usually contains menu items, radio groups
 * and/or submenus.
 */
@Component({
  tag: 'vm-menu',
  styleUrl: 'menu.css',
  shadow: true,
})
export class Menu {
  private shouldFocusOnOpen = false;

  private submenus!: NodeListOf<HTMLVmMenuElement>;

  @Element() el!: HTMLVmMenuElement;

  @State() menuItems!: NodeListOf<HTMLVmMenuItemElement>;

  @Watch('menuItems')
  onMenuItemsChange() {
    this.vmMenuItemsChange.emit(this.menuItems);
  }

  @State() currFocusedMenuItem = 0;

  @Watch('currFocusedMenuItem')
  async onFocusedMenuItemChange() {
    const menuItem = await this.getFocusedMenuItem();
    this.vmFocusMenuItemChange.emit(menuItem);
  }

  /**
   * Whether the menu is open/visible.
   */
  @Prop({ mutable: true, reflect: true }) active = false;

  @Watch('active')
  onActiveChange() {
    if (this.active) {
      this.findMenuItems();
      this.findSubmenus();
    }

    this.active ? this.vmOpen.emit() : this.vmClose.emit();
  }

  /**
   * The `id` attribute of the menu.
   */
  @Prop() identifier!: string;

  /**
   * The `id` attribute value of the control responsible for opening/closing this menu.
   */
  @Prop() controller!: string;

  /**
   * Emitted when the menu is open/active.
   */
  @Event() vmOpen!: EventEmitter<void>;

  /**
   * Emitted when the menu has closed/is not active.
   */
  @Event() vmClose!: EventEmitter<void>;

  /**
   * Emitted when the menu items present changes.
   */
  @Event() vmMenuItemsChange!: EventEmitter<NodeListOf<HTMLVmMenuItemElement> | undefined>;

  /**
   * Emitted when the currently focused menu item changes.
   */
  @Event() vmFocusMenuItemChange!: EventEmitter<HTMLVmMenuItemElement | undefined>;

  constructor() {
    withComponentRegistry(this);
  }

  connectedCallback() {
    this.findMenuItems();
  }

  componentDidRender() {
    if (this.active && this.shouldFocusOnOpen) {
      writeTask(() => {
        this.el.focus();
        this.shouldFocusOnOpen = false;
      });
    }
  }

  /**
   * Returns the controller responsible for opening/closing this menu.
   */
  @Method()
  async getController() {
    return document.querySelector(`#${this.controller}`) as HTMLElement;
  }

  /**
   * Returns the currently focused menu item.
   */
  @Method()
  async getFocusedMenuItem() {
    return this.menuItems[this.currFocusedMenuItem];
  }

  /**
   * This should be called directly before opening the menu to set the keyboard focus on it. This
   * is a one-time operation and needs to be called everytime prior to opening the menu.
   */
  @Method()
  async focusOnOpen() {
    this.shouldFocusOnOpen = true;
  }

  private findMenuItems() {
    this.menuItems = document.querySelectorAll(
      buildNoAncestorSelector(`#${this.identifier}`, 'vm-menu', 'vm-menu-item', 5),
    );
  }

  private async focusController() {
    const controller = await this.getController();
    controller?.focus();
  }

  private focusMenuItem(index: number) {
    let boundIndex = (index >= 0) ? index : (this.menuItems.length - 1);
    if (boundIndex >= this.menuItems.length) boundIndex = 0;
    this.currFocusedMenuItem = boundIndex;
    this.menuItems[boundIndex]?.focus();
  }

  private openSubmenu() {
    const menuItem = this.menuItems[this.currFocusedMenuItem];
    if (isUndefined(menuItem)) return;
    menuItem!.click();
    writeTask(() => {
      const submenu = document.querySelector(`#${menuItem!.menu}`) as HTMLVmMenuElement;
      submenu?.focus();
    });
  }

  private onOpen(event: Event) {
    event.stopPropagation();
    this.findMenuItems();
    this.currFocusedMenuItem = 0;
    // Prevents forwarding click event that opened the menu to menu item.
    setTimeout(() => { this.menuItems[this.currFocusedMenuItem]?.focus(); }, 10);
    this.active = true;
  }

  private onClose() {
    this.currFocusedMenuItem = -1;
    this.active = false;
    this.focusController();
  }

  private onClick(event: Event) {
    event.stopPropagation();
  }

  private onKeyDown(event: KeyboardEvent) {
    if (!this.active || this.menuItems.length === 0) return;
    event.preventDefault();
    event.stopPropagation();
    switch (event.key) {
      case 'Escape':
        this.onClose();
        break;
      case 'ArrowDown':
      case 'Tab':
        this.focusMenuItem(this.currFocusedMenuItem + 1);
        break;
      case 'ArrowUp':
        this.focusMenuItem(this.currFocusedMenuItem - 1);
        break;
      case 'ArrowLeft':
        this.onClose();
        break;
      case 'ArrowRight':
      case 'Enter':
      case ' ':
        this.openSubmenu();
        break;
      case 'Home':
      case 'PageUp':
        this.focusMenuItem(0);
        break;
      case 'End':
      case 'PageDown':
        this.focusMenuItem(this.menuItems!.length - 1);
        break;
    }
  }

  private findSubmenus() {
    this.submenus = document.querySelectorAll(
      buildNoAncestorSelector(`#${this.identifier}`, 'vm-menu', 'vm-menu', 4),
    );
  }

  private isValidSubmenu(submenu: HTMLElement | null) {
    if (isNull(submenu)) return false;
    return !isUndefined(Array.from(this.submenus).find((menu) => menu.id === submenu!.id));
  }

  private toggleSubmenu(submenu: HTMLVmMenuElement, isActive: boolean) {
    if (!this.isValidSubmenu(submenu)) return;

    Array.from(this.menuItems)
      .filter((menuItem) => menuItem.identifier !== submenu.controller)
      .forEach((menuItem) => { menuItem.hidden = isActive; });

    submenu.active = isActive;
  }

  @Listen('vmOpen')
  onSubmenuOpen(event: CustomEvent<void>) {
    const submenu = event.target as HTMLVmMenuElement;
    this.toggleSubmenu(submenu, true);
  }

  @Listen('vmClose')
  onSubmenuClose(event: CustomEvent<void>) {
    const submenu = event.target as HTMLVmMenuElement;
    this.toggleSubmenu(submenu, false);
  }

  @Listen('click', { target: 'window' })
  onWindowClick() {
    if (this.active) this.active = false;
  }

  @Listen('keydown', { target: 'window' })
  onWindowKeyDown(event: KeyboardEvent) {
    if (this.active && (event.key === 'Escape')) this.onClose();
  }

  render() {
    return (
      <div
        id={this.identifier}
        class="menu"
        role="menu"
        tabindex="-1"
        aria-labelledby={this.controller}
        aria-hidden={!this.active ? 'true' : 'false'}
        onFocus={this.onOpen.bind(this)}
        onClick={this.onClick.bind(this)}
        onKeyDown={this.onKeyDown.bind(this)}
      >
        {/* Seems like without this div unnecessary re-renders keep happening. */}
        <div>
          <slot />
        </div>
      </div>
    );
  }
}
