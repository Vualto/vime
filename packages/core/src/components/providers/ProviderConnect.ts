import { getElement, writeTask } from '@stencil/core';
import { Disposal } from '../../utils/Disposal';
import { listen } from '../../utils/dom';
import { createStencilHook } from '../../utils/stencil';
import { StateChange } from '../core/player/PlayerDispatcher';
import { PlayerProps } from '../core/player/PlayerProps';
import { AdapterHost, MediaProviderAdapter } from './MediaProvider';
import { Provider } from './Provider';
import { PROVIDER_CHANGE_EVENT } from './ProviderDispatcher';
import { ProviderWritableProps, isProviderWritableProp } from './ProviderProps';

export const PROVIDER_CACHE_KEY = Symbol('vmProviderCache');
export const PROVIDER_CONNECT_EVENT = 'vmMediaProviderConnect';
export const PROVIDER_DISCONNECT_EVENT = 'vmMediaProviderDisconnect';

export type ProviderCache = Map<keyof ProviderWritableProps, any>;

export type ProviderConnectEventDetail = AdapterHost;

export interface ProviderHost extends ProviderWritableProps {
  [PROVIDER_CACHE_KEY]?: ProviderCache
  ready: boolean
  currentProvider?: Provider
  logger?: PlayerProps['logger']
  provider?: AdapterHost
  adapter?: MediaProviderAdapter
  onProviderDisconnect?: () => void
}

export function withProviderHost(connector: ProviderHost) {
  const el = getElement(connector);
  const disposal = new Disposal();

  const cache: ProviderCache = new Map();
  connector[PROVIDER_CACHE_KEY] = cache;

  function initCache() {
    (Object.keys(connector) as (keyof ProviderWritableProps)[]).forEach((prop) => {
      cache.set(prop, connector[prop]);
    });
  }

  function onDisconnect(e?: Event) {
    e?.stopImmediatePropagation();
    writeTask(async () => {
      connector.ready = false;
      connector.provider = undefined;
      connector.adapter = undefined;
      cache.clear();
      connector.onProviderDisconnect?.();
    });
  }

  function onConnect(e: CustomEvent<ProviderConnectEventDetail>) {
    e.stopImmediatePropagation();
    initCache();

    const host = getElement(e.detail) as AdapterHost;
    if (connector.provider === host) return;

    const name = host
      ?.nodeName
      .toLowerCase()
      .replace('vm-', '');

    writeTask(async () => {
      onDisconnect();
      connector.provider = host;
      connector.adapter = await host?.getAdapter();
      connector.currentProvider = Object.values(Provider)
        .find((provider) => name === provider);
    });
  }

  function onChange(e: CustomEvent<StateChange<ProviderWritableProps>>) {
    e.stopImmediatePropagation();
    const { by, prop, value } = e.detail;

    if (!isProviderWritableProp(prop)) {
      connector.logger?.warn(`${by.nodeName} tried to change \`${prop}\` but it is readonly.`);
      return;
    }

    writeTask(() => {
      cache.set(prop, value);
      (connector as any)[prop] = value;
    });
  }

  createStencilHook(connector, () => {
    disposal.add(listen(el, PROVIDER_CONNECT_EVENT, onConnect));
    disposal.add(listen(el, PROVIDER_DISCONNECT_EVENT, onDisconnect));
    disposal.add(listen(el, PROVIDER_CHANGE_EVENT, onChange));
  }, () => {
    disposal.empty();
    cache.clear();
  });
}

export function withProviderConnect(ref: any) {
  const host = getElement(ref) as AdapterHost;

  const buildEvent = (name: string) => new CustomEvent<ProviderConnectEventDetail>(name, {
    bubbles: true,
    composed: true,
    detail: host,
  });

  const connectEvent = buildEvent(PROVIDER_CONNECT_EVENT);
  const disconnectEvent = buildEvent(PROVIDER_DISCONNECT_EVENT);

  createStencilHook(ref, () => {
    host.dispatchEvent(connectEvent);
  }, () => {
    host.dispatchEvent(disconnectEvent);
  });
}
