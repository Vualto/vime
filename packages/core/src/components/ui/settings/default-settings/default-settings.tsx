import { h, Component, Prop } from '@stencil/core';
import { withPlayerContext } from '../../../core/player/withPlayerContext';
import { PlayerProps } from '../../../core/player/PlayerProps';
import { Disposal } from '../../../../utils/Disposal';
import { Dispatcher, createDispatcher } from '../../../core/player/PlayerDispatcher';
import { findPlayer } from '../../../core/player/findPlayer';
import { withComponentRegistry } from '../../../core/player/withComponentRegistry';

/**
 * @slot - Used to extend the settings with additional menu options (see `vm-submenu` or
 * `vm-menu-item`).
 */
@Component({
  tag: 'vm-default-settings',
  shadow: true,
  styleUrl: 'default-settings.css',
})
export class DefaultSettings {
  private textTracksDisposal = new Disposal();

  private dispatch!: Dispatcher;

  private player?: HTMLVmPlayerElement;

  private rateSubmenu: any;

  private qualitySubmenu: any;

  private captionsSubmenu: any;

  /**
   * Pins the settings to the defined position inside the video player. This has no effect when
   * the view is of type `audio`, it will always be `bottomRight`.
   */
  @Prop({
    reflect: true,
  }) pin: 'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight' = 'bottomRight';

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps['i18n'] = {};

  /**
   * @internal
   */
  @Prop() playbackReady: PlayerProps['playbackReady'] = false;

  /**
   * @internal
   */
  @Prop() playbackRate: PlayerProps['playbackRate'] = 1;

  /**
   * @internal
   */
  @Prop() playbackRates: PlayerProps['playbackRates'] = [1];

  /**
   * @internal
   */
  @Prop() playbackQuality?: PlayerProps['playbackQuality'];

  /**
   * @internal
   */
  @Prop() playbackQualities: PlayerProps['playbackQualities'] = [];

  constructor() {
    withComponentRegistry(this);
    withPlayerContext(this, [
      'i18n',
      'playbackReady',
      'playbackRate',
      'playbackRates',
      'playbackQuality',
      'playbackQualities',
    ]);
  }

  async connectedCallback() {
    this.player = await findPlayer(this);
    this.dispatch = createDispatcher(this);
  }

  componentWillRender() {
    if (!this.playbackReady) return undefined;

    return Promise.all([
      this.buildPlaybackRateSubmenu(),
      this.buildPlaybackQualitySubmenu(),
    ]);
  }

  disconnectedCallback() {
    this.player = undefined;
    this.textTracksDisposal.empty();
  }

  private onPlaybackRateSelect(event: Event) {
    const radio = event.target as HTMLVmMenuRadioElement;
    this.dispatch('playbackRate', parseFloat(radio.value));
  }

  private async buildPlaybackRateSubmenu() {
    const canSetPlaybackRate = await this.player?.canSetPlaybackRate();

    if (this.playbackRates.length === 1 || !canSetPlaybackRate) {
      this.rateSubmenu = (
        <vm-menu-item label={this.i18n.playbackRate} hint={this.i18n.normal} />
      );
      return;
    }

    const formatRate = (rate: number) => ((rate === 1) ? this.i18n.normal : `${rate}`);

    const radios = this.playbackRates.map((rate) => (
      <vm-menu-radio
        label={formatRate(rate)}
        value={`${rate}`}
      />
    ));

    this.rateSubmenu = (
      <vm-submenu label={this.i18n.playbackRate} hint={formatRate(this.playbackRate)}>
        <vm-menu-radio-group
          value={`${this.playbackRate}`}
          onVmCheck={this.onPlaybackRateSelect.bind(this)}
        >
          {radios}
        </vm-menu-radio-group>
      </vm-submenu>
    );
  }

  private onPlaybackQualitySelect(event: Event) {
    const radio = event.target as HTMLVmMenuRadioElement;
    this.dispatch('playbackQuality', radio.value);
  }

  private async buildPlaybackQualitySubmenu() {
    const canSetPlaybackQuality = await this.player?.canSetPlaybackQuality();

    if (this.playbackQualities.length === 0 || !canSetPlaybackQuality) {
      this.qualitySubmenu = (
        <vm-menu-item
          label={this.i18n.playbackQuality}
          hint={this.playbackQuality ?? this.i18n.auto}
        />
      );
      return;
    }

    // @TODO this doesn't account for audio qualities yet.
    const getBadge = (quality: string) => {
      const verticalPixels = parseInt(quality.slice(0, -1), 10);
      if (verticalPixels > 2160) return 'UHD';
      if (verticalPixels >= 1080) return 'HD';
      return undefined;
    };

    const radios = this.playbackQualities.map((quality) => (
      <vm-menu-radio
        label={quality}
        value={quality}
        badge={getBadge(quality)}
      />
    ));

    this.qualitySubmenu = (
      <vm-submenu label={this.i18n.playbackQuality} hint={this.playbackQuality}>
        <vm-menu-radio-group
          value={this.playbackQuality}
          onVmCheck={this.onPlaybackQualitySelect.bind(this)}
        >
          {radios}
        </vm-menu-radio-group>
      </vm-submenu>
    );
  }

  render() {
    return (
      <vm-settings pin={this.pin}>
        {this.rateSubmenu}
        {this.qualitySubmenu}
        {this.captionsSubmenu}
        <slot />
      </vm-settings>
    );
  }
}
