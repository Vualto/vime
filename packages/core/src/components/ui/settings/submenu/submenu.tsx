import { h, Component, Prop } from '@stencil/core';
import { isNull } from '../../../../utils/unit';
import { withComponentRegistry } from '../../../core/player/withComponentRegistry';

let idCount = 0;

/**
 * @slot - Used to pass in the body of the submenu which is usually a set of choices in the form
 * of a radio group (`vm-menu-radio-group`).
 */
@Component({
  tag: 'vm-submenu',
  shadow: true,
})
export class Submenu {
  private id!: string;

  /**
   * The title of the submenu.
   */
  @Prop() label!: string;

  /**
   * Whether the submenu should be displayed or not.
   */
  @Prop() hidden = false;

  /**
   * This can provide additional context about the current state of the submenu. For example, the
   * hint could be the currently selected option if the submenu contains a radio group.
   */
  @Prop() hint?: string;

  /**
   * Whether the submenu is open/closed.
   */
  @Prop({ mutable: true, reflect: true }) active = false;

  constructor() {
    withComponentRegistry(this);
  }

  connectedCallback() {
    this.genId();
  }

  private invalidEventTarget(event: CustomEvent<void>) {
    return isNull(event.target) || (event.target! as HTMLElement).id !== this.id;
  }

  private onOpen(event: CustomEvent<void>) {
    if (this.invalidEventTarget(event)) return;
    this.active = true;
  }

  private onClose(event: CustomEvent<void>) {
    if (this.invalidEventTarget(event)) return;
    this.active = false;
  }

  private genId() {
    idCount += 1;
    this.id = `vm-submenu-${idCount}`;
  }

  private getControllerId() {
    return `${this.id}-controller`;
  }

  render() {
    return (
      <div>
        <vm-menu-item
          identifier={this.getControllerId()}
          hidden={this.hidden}
          menu={this.id}
          label={this.label}
          hint={this.hint}
          expanded={this.active}
        />
        <vm-menu
          identifier={this.id}
          controller={this.getControllerId()}
          active={this.active}
          onVmOpen={this.onOpen.bind(this)}
          onVmClose={this.onClose.bind(this)}
        >
          <slot />
        </vm-menu>
      </div>
    );
  }
}
