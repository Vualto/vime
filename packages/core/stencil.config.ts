import { Config } from '@stencil/core';
import { svelteOutputTarget } from './targets/svelte';
import { reactOutputTarget } from './targets/react';
import { angularOutputTarget } from './targets/angular';
import { vueOutputTarget } from './targets/vue';
import { vueNextOutputTarget } from './targets/vue-next';

export const config: Config = {
  namespace: 'Vime',
  taskQueue: 'async',
  outputTargets: [
    reactOutputTarget(),
    angularOutputTarget(),
    vueOutputTarget(),
    vueNextOutputTarget(),
    svelteOutputTarget(),
    {
      type: 'dist',
      esmLoaderPath: '../loader',
    },
    {
      type: 'dist-custom-elements-bundle',
    },
  ],
  testing: {
    globals: {
      'ts-jest': {
        diagnostics: false,
      },
    },
    setupFilesAfterEnv: [
      './src/globals/jest.js',
    ],
    coverageDirectory: 'jest-coverage',
    // @ts-ignore
    watchPlugins: [
      'jest-watch-typeahead/filename',
      'jest-watch-typeahead/testname',
    ],
    testPathIgnorePatterns: [
      '/node_modules/',
      '/cypress/',
    ],
  },
  preamble: '(C) Vime https://vimejs.com - MIT License',
};
