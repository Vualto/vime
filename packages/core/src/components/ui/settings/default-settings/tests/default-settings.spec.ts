import { SpecPage } from '@stencil/core/testing';
import { DefaultSettings } from '../default-settings';
import { newUISpecPage } from '../../../ui/tests';
import { Settings } from '../../settings/settings';
import { MenuRadioGroup } from '../../menu-radio-group/menu-radio-group';
import { MenuRadio } from '../../menu-radio/menu-radio';
import { MenuItem } from '../../menu-item/menu-item';
import { Menu } from '../../menu/menu';
import { Submenu } from '../../submenu/submenu';
import { MockMediaProviderAdapter } from '../../../../providers/MediaProvider';

let page: SpecPage;
let player: HTMLVmPlayerElement;
let provider: HTMLVmFaketubeElement;

const getSettingsMenu = () => page.root!.querySelector('vm-settings');

const getController = (label: string) => page.root!
  .querySelector(`vm-menu-item[aria-label="${label}"]`)! as HTMLVmMenuItemElement;

const getPlaybackRateSubmenuController = () => getController('Playback Rate');
const getPlaybackQualitySubmenuController = () => getController('Playback Quality');

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [
      DefaultSettings, Settings, Menu,
      Submenu, MenuRadioGroup, MenuRadio,
      MenuItem,
    ],
    '<vm-default-settings>DefaultSlot</vm-default-settings>',
  ));

  player = page.root! as HTMLVmPlayerElement;
  provider = page.root!.querySelector('vm-faketube')!;

  // Trigger another render cycle for the submenus to be built.
  provider.dispatchChange('playbackReady', true);
  await page.waitForChanges();
});

it('should be structurally sound', () => {
  expect(page.root).toMatchSnapshot();
});

test('playback rate submenu should be structurally sound', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackRate!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackRates', [1, 2, 3]);
  await page.waitForChanges();
  getPlaybackRateSubmenuController().click();
  await page.waitForChanges();
  expect(getSettingsMenu()).toMatchSnapshot();
});

it('should change playback rate on radio selection', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackRate!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackRates', [1, 2, 3]);
  await page.waitForChanges();
  getPlaybackRateSubmenuController().click();
  await page.waitForChanges();
  const radio = page.root!.querySelector('vm-menu-item[aria-label="2"]')! as HTMLVmMenuItemElement;
  radio.click();
  await page.waitForChanges();
  expect(player.playbackRate).toEqual(2);
});

it('should update selected radio on playback rate change', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackRate!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackRates', [1, 2, 3]);
  await page.waitForChanges();
  getPlaybackRateSubmenuController().click();
  await page.waitForChanges();
  player.playbackRate = 3;
  await page.waitForChanges();
  const radio = page.root!.querySelector('vm-menu-item[aria-label="3"]')! as HTMLVmMenuItemElement;
  expect(radio.checked).toBeTruthy();
});

test('playback quality submenu should be structurally sound', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackQuality!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackQuality', '720p');
  provider.dispatchChange('playbackQualities', ['1080p', '720p', '480p']);
  await page.waitForChanges();
  getPlaybackQualitySubmenuController().click();
  await page.waitForChanges();
  expect(getSettingsMenu()).toMatchSnapshot();
});

it('should change playback quality on radio selection', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackQuality!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackQuality', '720p');
  provider.dispatchChange('playbackQualities', ['1080p', '720p', '480p']);
  await page.waitForChanges();
  getPlaybackQualitySubmenuController().click();
  await page.waitForChanges();
  const radio = page.root!.querySelector('vm-menu-item[aria-label="1080p"]')! as HTMLVmMenuItemElement;
  radio.click();
  await page.waitForChanges();
  expect(player.playbackQuality).toEqual('1080p');
});

it('should update selected radio on playback quality change', async () => {
  const adapter = await player.getAdapter() as MockMediaProviderAdapter;
  adapter.canSetPlaybackQuality!.mockReturnValue(Promise.resolve(true));
  provider.dispatchChange('playbackQuality', '720p');
  provider.dispatchChange('playbackQualities', ['1080p', '720p', '480p']);
  await page.waitForChanges();
  getPlaybackQualitySubmenuController().click();
  await page.waitForChanges();
  player.playbackQuality = '480p';
  await page.waitForChanges();
  const radio = page.root!.querySelector('vm-menu-item[aria-label="480p"]')! as HTMLVmMenuItemElement;
  expect(radio.checked).toBeTruthy();
});
