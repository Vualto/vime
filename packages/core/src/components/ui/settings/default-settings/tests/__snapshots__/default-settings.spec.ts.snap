// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playback quality submenu should be structurally sound 1`] = `
<vm-settings pin="bottomRight" style="bottom: 0px; right: 8px;">
  <!---->
  <vm-menu aria-hidden="true" aria-labelledby="" id="vime-settings-5" role="menu" tabindex="-1">
    <!---->
    <div>
      <vm-menu-item aria-hidden="false" aria-label="Playback Rate" class="notTouch" role="menuitem" tabindex="0">
        Playback Rate
        <span class="spacer"></span>
        <span class="hint">
          Normal
        </span>
      </vm-menu-item>
      <vm-submenu>
        <!---->
        <vm-menu-item aria-controls="vime-submenu-4" aria-expanded="false" aria-haspopup="true" aria-hidden="false" aria-label="Playback Quality" class="notTouch" id="vime-submenu-4-controller" role="menuitem" tabindex="0">
          Playback Quality
          <span class="spacer"></span>
          <span class="hint">
            720p
          </span>
          <span class="arrow right"></span>
        </vm-menu-item>
        <vm-menu aria-hidden="true" aria-labelledby="vime-submenu-4-controller" id="vime-submenu-4" role="menu" tabindex="-1">
          <!---->
          <div>
            <vm-menu-radio-group>
              <!---->
              <vm-menu-radio>
                <vm-menu-item aria-checked="false" aria-hidden="false" aria-label="1080p" class="notTouch" role="menuitemradio" tabindex="0">
                  <vm-icon href="#vime-checkmark"></vm-icon>
                  1080p
                  <span class="spacer"></span>
                  <span class="badge">
                    HD
                  </span>
                </vm-menu-item>
              </vm-menu-radio>
              <vm-menu-radio>
                <vm-menu-item aria-checked="true" aria-hidden="false" aria-label="720p" class="notTouch" role="menuitemradio" tabindex="0">
                  <vm-icon href="#vime-checkmark"></vm-icon>
                  720p
                </vm-menu-item>
              </vm-menu-radio>
              <vm-menu-radio>
                <vm-menu-item aria-checked="false" aria-hidden="false" aria-label="480p" class="notTouch" role="menuitemradio" tabindex="0">
                  <vm-icon href="#vime-checkmark"></vm-icon>
                  480p
                </vm-menu-item>
              </vm-menu-radio>
            </vm-menu-radio-group>
          </div>
        <//vm-menu>
      </vm-submenu>
      <vm-menu-item aria-hidden="false" aria-label="Subtitles/CC" class="notTouch" role="menuitem" tabindex="0">
        Subtitles/CC
        <span class="spacer"></span>
        <span class="hint">
          None
        </span>
      </vm-menu-item>
      DefaultSlot
    </div>
  <vm-menu>
</vm-settings>
`;

exports[`playback rate submenu should be structurally sound 1`] = `
<vm-settings pin="bottomRight" style="bottom: 0px; right: 8px;">
  <!---->
  <vm-menu aria-hidden="true" aria-labelledby="" id="vime-settings-2" role="menu" tabindex="-1">
    <!---->
    <div>
      <vm-submenu>
        <!---->
        <vm-menu-item aria-controls="vime-submenu-1" aria-expanded="false" aria-haspopup="true" aria-hidden="false" aria-label="Playback Rate" class="notTouch" id="vime-submenu-1-controller" role="menuitem" tabindex="0">
          Playback Rate
          <span class="spacer"></span>
          <span class="hint">
            Normal
          </span>
          <span class="arrow right"></span>
        </vm-menu-item>
        <vm-menu aria-hidden="true" aria-labelledby="vime-submenu-1-controller" id="vime-submenu-1" role="menu" tabindex="-1">
          <!---->
          <div>
            <vm-menu-radio-group>
              <!---->
              <vm-menu-radio>
                <vm-menu-item aria-checked="true" aria-hidden="false" aria-label="Normal" class="notTouch" role="menuitemradio" tabindex="0">
                  <vm-icon href="#vime-checkmark"></vm-icon>
                  Normal
                </vm-menu-item>
              </vm-menu-radio>
              <vm-menu-radio>
                <vm-menu-item aria-checked="false" aria-hidden="false" aria-label="2" class="notTouch" role="menuitemradio" tabindex="0">
                  <vm-icon href="#vime-checkmark"></vm-icon>
                  2
                </vm-menu-item>
              </vm-menu-radio>
              <vm-menu-radio>
                <vm-menu-item aria-checked="false" aria-hidden="false" aria-label="3" class="notTouch" role="menuitemradio" tabindex="0">
                  <vm-icon href="#vime-checkmark"></vm-icon>
                  3
                </vm-menu-item>
              </vm-menu-radio>
            </vm-menu-radio-group>
          </div>
        <///vm-menu>
      </vm-submenu>
      <vm-menu-item aria-hidden="false" aria-label="Playback Quality" class="notTouch" role="menuitem" tabindex="0">
        Playback Quality
        <span class="spacer"></span>
        <span class="hint">
          Auto
        </span>
      </vm-menu-item>
      <vm-menu-item aria-hidden="false" aria-label="Subtitles/CC" class="notTouch" role="menuitem" tabindex="0">
        Subtitles/CC
        <span class="spacer"></span>
        <span class="hint">
          None
        </span>
      </vm-menu-item>
      DefaultSlot
    </div>
  <vm-menu>
</vm-settings>
`;

exports[`should be structurally sound 1`] = `
<vm-player aria-busy="false" aria-hidden="false" aria-label="Video Player" id="vime-player-1" ready="" tabindex="0">
  <!---->
  <vm-faketube></vm-faketube>
  <vm-ui>
    <!---->
    <div>
      <vm-default-settings pin="bottomRight">
        <!---->
        <vm-settings pin="bottomRight" style="bottom: 0px; right: 8px;">
          <!---->
          <vm-menu aria-hidden="true" aria-labelledby="" id="vime-settings-1" role="menu" tabindex="-1">
            <!---->
            <div>
              <vm-menu-item aria-hidden="false" aria-label="Playback Rate" class="notTouch" role="menuitem" tabindex="0">
                Playback Rate
                <span class="spacer"></span>
                <span class="hint">
                  Normal
                </span>
              </vm-menu-item>
              <vm-menu-item aria-hidden="false" aria-label="Playback Quality" class="notTouch" role="menuitem" tabindex="0">
                Playback Quality
                <span class="spacer"></span>
                <span class="hint">
                  Auto
                </span>
              </vm-menu-item>
              <vm-menu-item aria-hidden="false" aria-label="Subtitles/CC" class="notTouch" role="menuitem" tabindex="0">
                Subtitles/CC
                <span class="spacer"></span>
                <span class="hint">
                  None
                </span>
              </vm-menu-item>
              DefaultSlot
            </div>
          <//vm-menu>
        </vm-settings>
      </vm-default-settings>
    </div>
  </vm-ui>
</vm-player>
`;
